[LLVM][MC][DecoderEmitter] Add support to specialize decoder per bitwidth

jurahul opened the pull request:
This change adds an option to specialize decoders per bitwidth, which can help reduce the (compiled) code size of the decoder code.

**Current state**:
Currently, the code generated by the decoder emitter consists of two key functions: `decodeInstruction` which is the entry point into the generated code and `decodeToMCInst` which is invoked when a decode op is reached while traversing through the decoder table. Both functions are templated on `InsnType` which is the raw instruction bits that are supplied to `decodeInstruction`.

Several backends call `decodeInstruction` with different `InsnType` types, leading to several template instantiations of these functions in the final code. As an example, AMDGPU instantiates this function with type `DecoderUInt128` type for decoding 96/128-bit instructions, `uint64_t` for decoding 64-bit instructions, and `uint32_t` for decoding 32-bit instructions. Since there is just one `decodeToMCInst` in the generated code, it has code that handles decoding for *all* instruction sizes. However, the decoders emitted for different instructions sizes rarely have any intersection with each other. That means, in the AMDGPU case, the instantiation with InsnType == DecoderUInt128 has decoder code for 32/64-bit instructions that is *never exercised*. Conversely, the instantiation with InsnType == uint64_t has decoder code for 128/96/32-bit instructions that is never exercised. This leads to unnecessary dead code in the generated disassembler binary (that the compiler cannot eliminate by itself).

**New state**:
With this change, we introduce an option `specialize-decoders-per-bitwidth`. Under this mode, the DecoderEmitter will generate several versions of `decodeToMCInst` function, one for each bitwidth. The code is still templated, but will require backends to specify, for each `InsnType` used, the bitwidth of the instruction that the type is used to represent using a type-trait `InsnBitWidth`. This will enable the templated code to choose the right variant of `decodeToMCInst`. Under this mode, a particular instantiation will only end up instantiating a single variant of `decodeToMCInst` generated and that will include only those decoders that are applicable to a single bitwidth, resulting in elimination of the code duplication through instantiation and a reduction in code size.

Additionally, under this mode, decoders are uniqued only within a given bitwidth (as opposed to across all bitwidths without this option), so the decoder index values assigned are smaller, and consume less bytes in their ULEB128 encoding. As a result, the generated decoder tables can also reduce in size.

Adopt this feature for the AMDGPU and RISCV backend. In a release build, this results in a net 55% reduction in the .text size of libLLVMAMDGPUDisassembler.so and a 5% reduction in the .rodata size. For RISCV, which today uses a single `uint64_t` type, this results in a 3.7% increase in code size (expected as we instantiate the code 3 times now).

Actual measured sizes are as follows:
```
Baseline commit: 72c04bb882ad70230bce309c3013d9cc2c99e9a7
Configuration: Ubuntu clang version 18.1.3, release build with asserts disabled.
 
AMDGPU        Before       After      Change
======================================================
.text         612327       275607     55% reduction
.rodata       369728       351336      5% reduction          

RISCV:
======================================================
.text          47407       49187      3.7% increase   
.rodata        35768       35839      0.1% increase
```

jurahul committed:
Cull decoders

jurahul commented:
Hi @s-barannikov and @topperc, this is the new simplified version of the earlier change to compile out decoders that are unused for a particular bitwidth. This approach of using a type-trait seems simpler and is much less code changes than before. Note that it includes both the per-bitwidth decoder specialization as well as numbering decoders for each bitwidth, yielding some savings on the size of the decoder tables as well.

I'll fill in the description soon to explain the motivation and changes.

s-barannikov commented on llvm/utils/TableGen/DecoderEmitter.cpp:
```diff
@@ -2528,59 +2535,114 @@ namespace {
+
+  // Bucket entries in the `EncMap` based on the instruction bitwidths if
+  // SpecializeDecodersPerBitwidth is enabled.
+  SmallVector<SmallVector<const EncMapTy::value_type *>> PerBitWidthEncMap;
```
Another idea is to turn EncMap into a map of maps, where the outer map has bitwidth key, and the innner has (ns, hwmode) key. Then guard `TableInfo.Decoders.clear()` and `emitDecodeFunction` by the option.

jurahul replied:
I applied the second suggestion and yeah, that gets rid of the extra bucket vector and the nested loop. Thanks

jurahul commented on llvm/lib/Target/RISCV/Disassembler/RISCVDisassembler.cpp:
```diff
@@ -558,7 +558,8 @@ static DecodeStatus decodeXqccmpRlistS0(MCInst &Inst, uint32_t Imm,
 
-static DecodeStatus decodeCSSPushPopchk(MCInst &Inst, uint32_t Insn,
+static DecodeStatus decodeCSSPushPopchk(MCInst &Inst,
+                                        const std::bitset<48> &Insn,
```
You're right. `decodeCSSPushPopchk` can use uint16_t and `decodeXTHeadMemPair` can use uint32_t. I did get compiler errors/warnings, maybe I misread them. Fixed now (after updating).

s-barannikov commented on llvm/include/llvm/Target/Target.td:
```diff
@@ -1158,6 +1158,13 @@ class InstrInfo {
+  // different bitwidths and defining `InsnBitWidth` template specialization for
+  // the `InsnType` types used. Some common specializations are already defined
+  // in MCDecoder.h.
+  bit SpecializeDecodersPerBitwidth = false;
```
It's not about InstrInfo. My point is that decoder options should be, well, *decoder* options. Otherwise we regenerate *all* `*.inc` files, all `*.cpp` files that include them, and relink `llc` when the option changes. Changing the option value on the command line would lead to recompiling YourTargetDisassembler.cpp only. It is also useful when you want to experiment with various options / combinations of thereof, and makes unittests simpler (since you no longer need to duplicate tests for each option value, you can pass it on the command like and use `--check-prefix`).

jurahul replied:
Sounds good, I don't have a strong preference here, so will go with a command line option.

s-barannikov commented on llvm/utils/TableGen/DecoderEmitter.cpp:
```diff
@@ -2514,58 +2530,91 @@ namespace {
+        "Cannot specialize decoders for variable length instuctions");
+
+  // Entries in `EncMap` are already sorted by bitwidth. So bucketing per
+  // bitwidth can be done on-the-fly as we iterate over tha map.
```
typo "tha"

jurahul commented on llvm/utils/TableGen/DecoderEmitter.cpp:
```diff
@@ -93,6 +94,15 @@ static cl::opt<bool> UseFnTableInDecodeToMCInst(
+    "specialize-decoders-per-bitwidth",
+    cl::desc("Specialize the generated `decodeToMCInst` function per bitwidth. "
+             "Can help reduce the code size. This requires use of different "
+             "`InsnType` for different bitwidths and defining `InsnBitWidth` "
```
Addressed

jurahul commented:
> > ```
> > template <typename T>
> > constexpr bool isSupportedInsnBitWidth(uint32_t N) { return false; }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > This will allow using the same C++ type for > 1 bitwidths. Then we would need to remove the current specializations in MCDecoder.h as they assume a specific usage and then backends can choose. Not sure if that generalization is needed.
> 
> LMK if this is something to pursue. I suspect it may not be too much effort to do this and offers additional flexibility for backends.

I did and now RISCV can continue using uint64_t for 48-bit insts. Will update the PR soon

jurahul commented:
> The windows build failure seems to be this : https://stackoverflow.com/questions/29502052/template-specialization-and-enable-if-problems

The specific form that seems to work with MSVC (and clang and gcc) was derived here: https://godbolt.org/z/Eqdf4Tehz

jurahul commented:
Looks like MSVC builds with compiler version 14.x do not support either form, so maybe not possible to pursue this at the moment.

jurahul commented on llvm/include/llvm/MC/MCDecoder.h:
```diff
@@ -72,20 +72,17 @@ insertBits(IntType &field, IntType bits, unsigned startBit, unsigned numBits) {
+// types, only a single one of them should return `isSupportedInsnBitWidth` as
+// true for a given bit width.
+template <typename T>
+inline constexpr bool isSupportedInsnBitWidth(uint32_t BitWidth) {
```
Thanks. Given the MSVC troubles, I think I am going to revert back to the earlier stage of just support one bitwidth per instruction, but without proving any default implementation, so RISCV can still choose to use uint64_t for 48 bit insts.

jurahul commented:
@s-barannikov this is ready for another round of review. The delta is that there are no longer some default implementations of `InsnBitWidth` in MCDecoder.h, so individual targets need to provide them according to their usage. This enables, for example, for RISCV to use uint64_t for 48 bit instructions.

jurahul commented on llvm/lib/Target/RISCV/Disassembler/RISCVDisassembler.cpp:
```diff
@@ -700,6 +700,12 @@ static constexpr DecoderListEntry DecoderList32[]{
+template <> inline constexpr uint32_t llvm::MCD::InsnBitWidth<uint16_t> = 16;
+template <> inline constexpr uint32_t llvm::MCD::InsnBitWidth<uint32_t> = 32;
+// Use uint64_t to represent 48 bit instructions.
+template <> inline constexpr uint32_t llvm::MCD::InsnBitWidth<uint64_t> = 48;
```
You cannot. This is what I attempted earlier with isSupportedBitWidth() and found that it does not work with all MSVC versions we support.

jurahul commented on llvm/lib/Target/RISCV/Disassembler/RISCVDisassembler.cpp:
```diff
@@ -700,6 +700,12 @@ static constexpr DecoderListEntry DecoderList32[]{
+template <> inline constexpr uint32_t llvm::MCD::InsnBitWidth<uint16_t> = 16;
+template <> inline constexpr uint32_t llvm::MCD::InsnBitWidth<uint32_t> = 32;
+// Use uint64_t to represent 48 bit instructions.
+template <> inline constexpr uint32_t llvm::MCD::InsnBitWidth<uint64_t> = 48;
```
static?

s-barannikov commented on llvm/include/llvm/MC/MCDecoder.h:
```diff
@@ -62,6 +72,13 @@ insertBits(IntType &field, IntType bits, unsigned startBit, unsigned numBits) {
+// it an invalid type for use as `InsnType` when instantiating the decoder.
+// Individual targets are expected to provide specializations for these based
+// on their usage.
+template <typename T> inline constexpr uint32_t InsnBitWidth = 0;
```
This one *could* be `static` as well (unlike specializations, it should have the same value in all translation units, so `inline` works here).
Perhaps it *should* be made `static` anyway to prevent accidental copy&paste bugs when specializing it.

s-barannikov reviewed (approved):
LGTM, but please for @topperc review

Meanwhile, a couple of tests wouldn't hurt.

s-barannikov commented:
The description needs to be updated

jurahul commented:
Yes thanks. Will add tests and update the description with one final set of numbers.

jurahul commented:
Updated description and added a unit test (which found that this option did not work with `use-fn-table-in-decode-to-mcinst`, so added additional fixes for that, and verified that check-llvm-mc-disassembler-{amdgup,riscv} pass locally with both options enabled).

jurahul commented:
@topperc can you please review as well?

jurahul commented:
Gentle ping @topperc, would like to get this in early next week (assuming review is ok).

s-barannikov commented on llvm/utils/TableGen/DecoderEmitter.cpp:
```diff
@@ -2514,58 +2548,91 @@ namespace {
     for (unsigned EncodingID : EncodingIDs) {
       const InstructionEncoding &Encoding = Encodings[EncodingID];
       const Record *EncodingDef = Encoding.getRecord();
       unsigned Size = EncodingDef->getValueAsInt("Size");
+      const unsigned BitWidth = IsVarLenInst ? MaxInstLen : 8 * Size;
```
```suggestion
      unsigned BitWidth = IsVarLenInst ? MaxInstLen : Encoding.getBitWidth();
```
and remove unused Size.

jurahul replied:
thanks, done

jrtc27 commented on llvm/lib/Target/RISCV/Disassembler/RISCVDisassembler.cpp:
```diff
@@ -700,6 +700,12 @@ static constexpr DecoderListEntry DecoderList32[]{
+template <> static constexpr uint32_t llvm::MCD::InsnBitWidth<uint16_t> = 16;
+template <> static constexpr uint32_t llvm::MCD::InsnBitWidth<uint32_t> = 32;
+// Use uint64_t to represent 48 bit instructions.
+template <> static constexpr uint32_t llvm::MCD::InsnBitWidth<uint64_t> = 48;
```
Yeah this seems backwards. Shouldn't the decoder know what bit width it's for rather than trying to reinfer it from the type that ended up being used to represent that bit width?

jurahul replied:
@jrtc27 The decoder know the number of bits, but not the actual C++ type to be used for that bitwidth (it generates templated code and backends are free to use whatever compatible types they want to use). `InsnBitWidth` is a way for the particular target to communicate this intent to the decoder generated code.

jurahul replied:
A previous iteration of this was trying to enforce the actual type as well, but that turned out to be much larger change (see https://github.com/llvm/llvm-project/pull/146593).

jurahul replied:
So something like?

```
template <>
struct TypeForBitWidth<32> {
  using T = uint32_t;
};
template <>
struct TypeForBitWidth<64> {
  using T = uint64_t;
};

int decodeToMCInst(TypeForBitWidth<32>::T Insn) {
  return 32;
}

int decodeToMCInst(TypeForBitWidth<64>::T Insn) {
  return 64;
}

int main() {
  uint32_t x = 0;
  std::cout << decodeToMCInst(x) << '\n';
  uint64_t y = 0;
  std::cout << decodeToMCInst(y) << '\n';
}
```
@topperc and @s-barannikov WDYT? It conveys the same information, but I agree its more directly expressing things. Note that with this though it won't directly work if/when we want to support > 1 bitwidth for a given type (which I did attempt in this PR but gave up because MSVC broke).

topperc replied:
I'm not sure overloading on type is a good idea. For this reason and calling the function with the wrong type issue raised here https://github.com/llvm/llvm-project/pull/154865#issuecomment-3225069066

jurahul replied:
Sounds good. I propose we keep the current scheme for now,

jurahul committed:
Review feedback

jurahul commented on llvm/utils/TableGen/DecoderEmitter.cpp:
```diff
@@ -2499,57 +2533,90 @@ namespace {
           Encoding.getBitWidth();
     }
+
+    // For variable instruction, we emit a instruction length table to let the
```
Fixed

topperc reviewed (approved):
LGTM. I guess we'll find out if there are any out of tree RISC-V vendors with 64 bit instructions.

s-barannikov commented on llvm/utils/TableGen/DecoderEmitter.cpp:
```diff
@@ -2499,57 +2533,90 @@ namespace {
   for (const auto &[HwModeID, EncodingIDs] : EncodingIDsByHwMode) {
     for (unsigned EncodingID : EncodingIDs) {
       const InstructionEncoding &Encoding = Encodings[EncodingID];
       const Record *EncodingDef = Encoding.getRecord();
```
Dead variable

jurahul replied:
Fixed.

jurahul committed:
Drop unused variable

s-barannikov commented:
Heads up, pre-merge checks seem to be broken:
```
2025-09-01T21:09:25.5173264Z /home/gha/actions-runner/_work/llvm-project/llvm-project/llvm/lib/Target/RISCV/Disassembler/RISCVDisassembler.cpp:705:13: error: explicit specialization cannot have a storage class [-Werror,-Wexplicit-specialization-storage-class]
2025-09-01T21:09:25.5174352Z   705 | template <> static constexpr uint32_t llvm::MCD::InsnBitWidth<uint16_t> = 16;
2025-09-01T21:09:25.5174684Z       |             ^~~~~~
2025-09-01T21:09:25.5175620Z /home/gha/actions-runner/_work/llvm-project/llvm-project/llvm/lib/Target/RISCV/Disassembler/RISCVDisassembler.cpp:706:13: error: explicit specialization cannot have a storage class [-Werror,-Wexplicit-specialization-storage-class]
2025-09-01T21:09:25.5176769Z   706 | template <> static constexpr uint32_t llvm::MCD::InsnBitWidth<uint32_t> = 32;
2025-09-01T21:09:25.5177142Z       |             ^~~~~~
2025-09-01T21:09:25.5177958Z /home/gha/actions-runner/_work/llvm-project/llvm-project/llvm/lib/Target/RISCV/Disassembler/RISCVDisassembler.cpp:708:13: error: explicit specialization cannot have a storage class [-Werror,-Wexplicit-specialization-storage-class]
2025-09-01T21:09:25.5178922Z   708 | template <> static constexpr uint32_t llvm::MCD::InsnBitWidth<uint64_t> = 48;
2025-09-01T21:09:25.5179256Z       |             ^~~~~~
2025-09-01T21:09:25.5179438Z 3 errors generated.
```
Not sure how this PR did pass the checks (maybe not all bots use -Werror?).

jurahul commented:
Thanks, looks like we need to just drop the `static` from these? I am testing locally and will create a PR.

s-barannikov commented:
> Thanks, looks like we need to just drop the `static` from these? I am testing locally and will create a PR.

That's what clang suggests me doing, I'm not really good with templates.

jurahul commented:
BTW, where did you see the message?

s-barannikov commented:
https://github.com/llvm/llvm-project/actions/runs/17387167049/job/49355461021?pr=156358
[raw logs](https://productionresultssa7.blob.core.windows.net/actions-results/efd42412-673a-4b5c-99f4-38f58012ff25/workflow-job-run-7bb6fdd7-fa80-51c3-8154-cc6145503c3c/logs/job/job-logs.txt?rsct=text%2Fplain&se=2025-09-01T21%3A50%3A51Z&sig=rkcSGEY5kHwjubxQEp6mtusLMXfxatGIN6Yb8efe0vQ%3D&ske=2025-09-02T09%3A32%3A23Z&skoid=ca7593d4-ee42-46cd-af88-8b886a2f84eb&sks=b&skt=2025-09-01T21%3A32%3A23Z&sktid=398a6654-997b-47e9-b12b-9515b896b4de&skv=2025-05-05&sp=r&spr=https&sr=b&st=2025-09-01T21%3A40%3A46Z&sv=2025-05-05)

nico commented:
Do these errors make sense to you? http://45.33.8.238/macm1/113320/step_3.txt

(That bot uses the GN build; might be a bug in the GN bug.)

jurahul commented:
> Do these errors make sense to you? http://45.33.8.238/macm1/113320/step_3.txt
> 
> (That bot uses the GN build; might be a bug in the GN build.)

Hmm..They don't. This seems different from the issue that @s-barannikov reported above. It looks like the .inc file is a dump of the records in the .td files, meaning that the new `-specialize-decoders-per-bitwidth` option is likely not getting recognized and then tablegen is just dumping all the class records to the output. I suspect something like that is happening. I don't see llvm-tblgen being rebuilt in the short log provided.

jurahul commented:
One the topic of using the same underlying type for > 1 bitwidths: I had earlier prototyped using something like:

```
tempate <typename T> constexpr bool isBitwidthSupported(uint32_t N) { return false;}
template<> constexpr bool isBitwidthSupported<uint64_t>(uint32_t N) { return N == 48 || N == 64; }
```

This ran into MSVC issues and I wasn't able to find a code pattern that worked. However, it seems we can achieve the same by having a wrapper type to do this:

```
template <typename T, size_ N> // represents a N bit int using an underlying type `T`.
class UIntNBits {
  T value;
  UIntNBits(T v) : value(v) {
     static_assert(std::is_integral_v<T>, "Underlying type expected to be int");
  }
  uint64_t extractBitsAsZExtValue() {  ... }
};
```

We can do this for RISCV if we do need 48 vs 64-bit distinction and do not want to use `std::bitset<>`.  I am not sure through if `std::bitset<48>` is any different, will need to look at the generated code for one platform and see if there is any difference.
